src/client/pythonEnvironments/collection/environmentsReducer.ts

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { Event, EventEmitter } from 'vscode';
+import { PythonEnvInfo } from '../base/info';
+import {
+    ILocator, IPythonEnvsIterator, PythonEnvUpdatedEvent, QueryForEvent,
+} from '../base/locator';
+import { PythonEnvsChangedEvent } from '../base/watcher';
+import { mergeEnvironments } from '../info';
+
+type PythonEnvsIteratorState = {
+    /**
+     * Set to true when all information from incoming iterator has been received
+     */
+    done: boolean;
+    /**
+     * Carries the number of pending background calls ongoing at the moment
+     */
+    pending: number;
+};
+
+export class PythonEnvsReducer implements ILocator {
+    public get onChanged(): Event<PythonEnvsChangedEvent> {
+        return this.pythonEnvsManager.onChanged;
+    }
+
+    constructor(private readonly pythonEnvsManager: ILocator) {}
+
+    public resolveEnv(env: string | PythonEnvInfo): Promise<PythonEnvInfo | undefined> {
+        return this.pythonEnvsManager.resolveEnv(env);
+    }
+
+    public iterEnvs(query?: QueryForEvent<PythonEnvsChangedEvent>): IPythonEnvsIterator {
+        const didUpdate = new EventEmitter<PythonEnvUpdatedEvent | null>();
+        return { ...this.iterEnvsIterator(didUpdate, query), onUpdated: didUpdate.event };
+    }
+
+    private async* iterEnvsIterator(
+        didUpdate: EventEmitter<PythonEnvUpdatedEvent | null>,
+        query?: QueryForEvent<PythonEnvsChangedEvent>,
+    ): AsyncIterator<PythonEnvInfo, void> {
+        const state: PythonEnvsIteratorState = {
+            done: false,
+            pending: 0,
+        };
+        const seen: PythonEnvInfo[] = [];
+        const iterator = this.pythonEnvsManager.iterEnvs(query);
+
+        if (iterator.onUpdated !== undefined) {
+            iterator.onUpdated((event) => {
+                if (event === null) {
+                    state.done = true;
+                    checkIfFinishedAndNotify(state, didUpdate);
+                } else {
+                    const old = seen.find((s) => s.location === event.old.location);
+                    if (old !== undefined) {
+                        state.pending += 1;
+                        resolveDifferencesInBackground(old, event.new, state, didUpdate).ignoreErrors();
+                    }
+                }
+            });
+        }
+
+        let result = await iterator.next();
+        while (!result.done) {
+            const currEnv = result.value;
+            const old = seen.find((s) => s.location === currEnv.location);

```

Comment:
likewise with the check:

```suggestion
            const old = seen.find((s) => isSameEnv(s, currEnv.location));
```

Reply:
As pointed out in other comments, Karthik's working on the same env method. I'll add that after he's done.